File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_registry.rb
require 'cucumber/cucumber_expressions/parameter'

module Cucumber
  module CucumberExpressions
    class ParameterRegistry
      INTEGER_REGEXPS = [/-?\d+/, /\d+/]
      FLOAT_REGEXP = /-?\d*\.?\d+/

      def initialize
        @parameters_by_type_name = {}
        @parameters_by_capture_group_regexp = {}
        @parameters_by_class = {}

        add_predefined_parameter(Parameter.new('int', Integer, INTEGER_REGEXPS, lambda { |s| s.to_i }))
        add_predefined_parameter(Parameter.new('float', Float, FLOAT_REGEXP, lambda { |s| s.to_f }))
      end

      def lookup_by_type(type)
        if type.is_a?(Class)
          lookup_by_class(type)
        elsif type.is_a?(String)
          lookup_by_type_name(type)
        else
          raise Exception.new("Type must be string or class, but was #{type} of type #{type.class}")
        end
      end

      def lookup_by_class(clazz)
        parameter = @parameters_by_class[clazz]
        if parameter.nil?
          create_anonymous_lookup(lambda { |s| clazz.new(s) })
        else
          parameter
        end
      end

      def lookup_by_type_name(type_name)
        @parameters_by_type_name[type_name]
      end

      def lookup_by_capture_group_regexp(capture_group_regexp)
        @parameters_by_capture_group_regexp[capture_group_regexp]
      end

      def create_anonymous_lookup(proc)
        Parameter.new(nil, nil, ['.+'], proc)
      end

      def parameters
        @parameters_by_type_name.values
      end

      def add_parameter(parameter)
        add_parameter0(parameter, true)
      end

      private

      def add_predefined_parameter(parameter)
        add_parameter0(parameter, false)
      end

      def add_parameter0(parameter, check_conflicts)
        put(@parameters_by_type_name, parameter.type_name, parameter, "type name", check_conflicts)
        put(@parameters_by_class, parameter.type, parameter, "type", check_conflicts)

        parameter.capture_group_regexps.each do |capture_group_regexp|
          put(@parameters_by_capture_group_regexp, capture_group_regexp, parameter, "regexp", check_conflicts)
        end
      end

      def put(map, key, parameter, prop, check_conflicts)
        if check_conflicts && map.has_key?(key)
          raise "There is already a parameter with #{prop} #{key}"
        end
        map[key] = parameter
      end
    end
  end
end

File: cucumber-expressions/javascript/src/parameter_registry.js
const Parameter = require('./parameter')

class ParameterRegistry {
  constructor() {
    this._parametersByTypeName = new Map()
    this._parametersByCaptureGroupRegexp = new Map()
    this._parametersByConstructorName = new Map()

    const INTEGER_REGEXPS = [/-?\d+/, /\d+/]
    const FLOAT_REGEXP = /-?\d*\.?\d+/

    this._addPredefinedParameter(new Parameter('int', Number, INTEGER_REGEXPS, parseInt))
    this._addPredefinedParameter(new Parameter('float', Number, FLOAT_REGEXP, parseFloat))
  }

  get parameters() {
    return this._parametersByTypeName.values()
  }

  lookupByType(type) {
    if (typeof type === 'function') {
      return this.lookupByFunction(type)
    } else if (typeof type === 'string') {
      return this.lookupByTypeName(type)
    } else {
      throw new Error(`Type must be string or function, but was ${type} of type ${typeof type}`)
    }
  }

  lookupByFunction(fn) {
    if (fn.name) {
      const prefix = fn.name[0]
      const looksLikeConstructor = prefix.toUpperCase() === prefix

      let parameter
      if (looksLikeConstructor) {
        parameter = this._parametersByConstructorName.get(fn.name)
      }
      if (!parameter) {
        const factory = s => {
          if (looksLikeConstructor) {
            return new fn(s)
          } else {
            return fn(s)
          }
        }
        return this.createAnonymousLookup(factory)
      } else {
        return parameter
      }
    } else {
      return this.createAnonymousLookup(fn)
    }
  }

  lookupByTypeName(typeName) {
    return this._parametersByTypeName.get(typeName)
  }

  lookupByCaptureGroupRegexp(captureGroupRegexp) {
    return this._parametersByCaptureGroupRegexp.get(captureGroupRegexp)
  }

  createAnonymousLookup(fn) {
    return new Parameter(null, null, [".+"], fn)
  }

  addParameter(parameter) {
    this._addParameter(parameter, true)
  }

  _addPredefinedParameter(parameter) {
    this._addParameter(parameter, false)
  }

  _addParameter(parameter, checkConflicts) {
    set(this._parametersByConstructorName, parameter.constructorFunction.name, parameter, 'constructor', checkConflicts)
    set(this._parametersByTypeName, parameter.typeName, parameter, 'type name', checkConflicts)

    for (let captureGroupRegexp of parameter.captureGroupRegexps) {
      set(this._parametersByCaptureGroupRegexp, captureGroupRegexp, parameter, 'regexp', checkConflicts)
    }
  }
}

function set(map, key, value, prop, checkConflicts) {
  if(checkConflicts && map.has(key))
    throw new Error(`There is already a parameter with ${prop} ${key}`)
  map.set(key, value)
}

module.exports = ParameterRegistry
